multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/clients/ApplicationRoutesGetter.java

```diff
@@ -16,24 +18,64 @@ public ApplicationRoutesGetter(CloudControllerClient client) {
         super(client);
     }
 
-    public List<CloudRoute> getRoutes(String appName) {
+    public List<CloudRouteExtended> getRoutes(String appName) {
         return new CustomControllerClientErrorHandler().handleErrorsOrReturnResult(() -> getRoutesInternal(appName));
     }
 
-    private List<CloudRoute> getRoutesInternal(String appName) {
+    private List<CloudRouteExtended> getRoutesInternal(String appName) {
         String appRoutesUrl = getAppRoutesUrl(client.getApplicationGuid(appName));
-        List<Map<String, Object>> resources = getAllResources(appRoutesUrl);
-        return toCloudRoutes(resources);
+        var resourcesWithIncluded = getAllResourcesWithIncluded(appRoutesUrl);
+        String routeGuids = getRouteGuidsAsString(resourcesWithIncluded.getResources());
+        String serviceRouteBindingsUrl = getServiceRouteBindingsUrl(routeGuids);
+        var serviceRouteBindingResources = getAllResources(serviceRouteBindingsUrl);
+        var serviceRouteBindingsByRouteGuid = groupByRouteGuid(serviceRouteBindingResources);
+
+        return toCloudRoutes(resourcesWithIncluded, serviceRouteBindingsByRouteGuid);
     }
 
     private String getAppRoutesUrl(UUID appGuid) {
-        return "/v2/apps/" + appGuid.toString() + "/routes?inline-relations-depth=1";
+        return "/v3/routes?include=domain&app_guids=" + appGuid.toString();
+    }
+
+    private String getServiceRouteBindingsUrl(String routeGuids) {
+        return "/v3/service_route_bindings?route_guids=" + routeGuids;
+    }
+
+    private String getRouteGuidsAsString(List<Map<String, Object>> routeResources) {
+        return routeResources.stream()
```

Comment:
We have to add check for length of route guids because there is length limit for url/query parameters. We have already such logic for batching in CloudControllerRestClientImp class -> `getApplicationResourcesByNamesInBatches`